pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/ast/ModelASTAgent.java

```diff
@@ -1,65 +1,107 @@
+            o.accumulate("key", entry.getKey().toJSON());
+            o.accumulate("value", entry.getValue().toJSON());
+            a.add(o);
+        }
+        return a;
+
     }
 
     @Override
     public void validate(ModelValidator validator) {
         validator.validateElement(this);
 
-        args.validate(validator);
+        for (Map.Entry<ModelASTKey, ModelASTValue> entry : variables.entrySet()) {
+            entry.getKey().validate(validator);
+            entry.getValue().validate(validator);
+        }
+    }
+
+    public List<String> getKeyNames() {
+        List<String> names = new ArrayList<>();
+        for (ModelASTKey k : variables.keySet()) {
+            names.add(k.getKey());
+        }
+        return names;
+    }
+
+    public ModelASTKey keyFromName(String name) {
+        for (ModelASTKey k : variables.keySet()) {
+            if (name.equals(k.getKey())) {
+                return k;
+            }
+        }
+        return null;
     }
 
     @Override
     public String toGroovy() {
-        String argStr;
+        StringBuilder argStr = new StringBuilder();
         // TODO: Stop special-casing agent none.
-        if (args instanceof ModelASTSingleArgument && (
-                ((ModelASTSingleArgument) args).getValue().getValue().equals("none") || ((ModelASTSingleArgument) args)
-                        .getValue().getValue().equals("any"))) {
-            argStr = (String)((ModelASTSingleArgument) args).getValue().getValue();
+        List<String> keys = getKeyNames();
+        if (keys.size() == 1 && (keys.contains("none") || keys.contains("any"))) {
+            argStr.append(keys.get(0));
         } else {
-            argStr = args.toGroovy();
+            argStr.append("{\n");
+            for (Map.Entry<ModelASTKey, ModelASTValue> entry: variables.entrySet()) {
+                argStr.append(entry.getKey().toGroovy()).append(" ").append(entry.getValue().toGroovy()).append("\n");
+            }
+            argStr.append("}");
         }
 
-        return "agent " + argStr + "\n";
+        return "agent " + argStr.toString() + "\n";
     }
 
     @Override
     public void removeSourceLocation() {
         super.removeSourceLocation();
-        args.removeSourceLocation();
+        for (Map.Entry<ModelASTKey, ModelASTValue> entry : variables.entrySet()) {
+            entry.getKey().removeSourceLocation();
+            entry.getValue().removeSourceLocation();
+        }
     }
 
-    public ModelASTArgumentList getArgs() {
-        return args;
+    public Map<ModelASTKey, ModelASTValue> getVariables() {
+        return variables;
     }
 
-    public void setArgs(ModelASTArgumentList args) {
-        this.args = args;
+    public void setVariables(Map<ModelASTKey, ModelASTValue> variables) {
+        this.variables = variables;
     }
 
     @Override
     public String toString() {
         return "ModelASTAgent{" +
-                "args=" + args +
+                "variables=" + variables +
```

Comment:
we've got the same thing in `ModelASTEnvironment` and elsewhere - I'll poke at it.